### PR TITLE
[SDK][Internal] Set context for experiment

### PR DESCRIPTION
**Pull request by brynn-code**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py [hunk 1]
```diff
@@ -45,6 +47,20 @@ def __init__(self, client):
...
+
+        :param experiment: Experiment name or experiment template absolute path.
+        :type experiment: str
+        :param test: Whether it's a test run. If True, will set line run id.
+        :type test: bool
+        """
+        operation_context = OperationContext.get_instance()
+        operation_context.experiment = experiment
+        if test:
+            operation_context.line_run_id = uuid.uuid4()
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py:62]: for `line_run_id`, this should be different across different flow/node

**Comment by brynn-code** [comment]
How to add e2e test?

File: src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py [hunk 2]
```diff
@@ -45,6 +47,31 @@ def __init__(self, client):
...
         self.command_submitter = ExperimentCommandSubmitter(self.run_operations)
 
+    @classmethod
+    def _set_experiment(cls, experiment):
+        """Set experiment info to operation context.
+
+        :param experiment: Experiment name or experiment template absolute path.
+        :type experiment: str
+        """
+        operation_context = OperationContext.get_instance()
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py:57]: add TODO that we should pass through the operation_context when invoke flow or script node cross process.

File: src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py [hunk 3]
```diff
@@ -45,6 +47,31 @@ def __init__(self, client):
...
+        operation_context._add_otel_attributes("experiment", experiment)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py:58]: Let's move _add_otel_attributes to start_trace function. Pass This info via a json objec in environment variable

File: src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py [hunk 4]
```diff
@@ -45,6 +47,31 @@ def __init__(self, client):
...
+        :param node_name: Experiment node name.
+        :type node_name: str
+        :param test_context: Experiment test context.
+        :type test_context: ~promptflow._sdk._submitter.experiment_orchestrator.ExperimentTemplateTestContext
+        """
+        operation_context = OperationContext.get_instance()
+        referenced_ids = test_context.node_name_to_referenced_id.get(node_name, [])
+        if referenced_ids:
+            # Here we use next to get the first referenced id, as we only support 1 referenced node for flow node now.
+            operation_context._add_otel_attributes("referenced.line_run_id", next(iter(referenced_ids)))
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py:73]: we can also do this the same way as experiment

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 5]
```diff
@@ -1065,6 +1065,7 @@ def execute_flow(
     run_aggregation: bool = True,
     enable_stream_output: bool = False,
     allow_generator_output: bool = False,  # TODO: remove this
+    run_id: str = None,
```
Comment (wangchao1230) [inline src/promptflow/promptflow/executor/flow_executor.py:1068]: I think we might need another sync on this

File: src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py (hunk 2 above)
Comment (brynn-code) [reply src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py]: removed operation context

**Review by zhengfeiwang — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Review by zhengfeiwang — APPROVED** [review]

**Review by wangchao1230 — DISMISSED** [review]

**Review by zhengfeiwang — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

**Pull request merged**